auth0/v3/authentication/token_verifier.py

```diff
@@ -0,0 +1,225 @@
+        keys = self._fetch_jwks()
+
+        if keys and key_id in keys:
+            return keys[key_id]
+
+        if not self._cache_is_fresh:
+            keys = self._fetch_jwks(force=True)
+            if keys and key_id in keys:
+                return keys[key_id]
+        raise TokenValidationError('RSA Public Key with ID "{}" was not found.'.format(key_id))
+
+
+class TokenVerifier():
+
+    def __init__(self, signature_verifier, issuer, audience, leeway=0, _clock=None):
+        if not signature_verifier or not isinstance(signature_verifier, SignatureVerifier):
+            raise TypeError("signature_verifier must be an instance of SignatureVerifier.")
+
+        self._options = {
+            'signature_verifier': signature_verifier,
+            'iss': issuer,
+            'aud': audience,
+            'leeway': leeway,
+            '_clock': _clock
+        }
+
+    def verify(self, token, nonce=None, max_age=None):
+        # Verify token presence
+        if not token or type(token) != str:
+            raise TokenValidationError("ID token is required but missing.")
+        if nonce and type(nonce) != str:
+            raise TokenValidationError("Nonce not specified.")
+        if max_age and type(max_age) != int:
+            raise TokenValidationError("Max Age not specified.")
+
+        opt = self._options.copy()
+        opt['nonce'] = nonce
+        opt['max_age'] = max_age
+
+        # Verify algorithm and signature
+        payload = opt['signature_verifier'].verify_signature(token)
+
+        # Verify claims
+        # Issuer
+
+        if 'iss' not in payload or type(payload['iss']) != str:
+            raise TokenValidationError('Issuer (iss) claim must be a string present in the ID token')
+        if payload['iss'] != opt['iss']:
+            raise TokenValidationError('Issuer (iss) claim mismatch in the ID token; expected "{}", found "{}"'.format(opt['iss'], payload['iss']))
+
+        # Subject
+        if 'sub' not in payload or type(payload['sub']) != str:
+            raise TokenValidationError('Subject (sub) claim must be a string present in the ID token')
+
+        # Audience
+        if 'aud' not in payload or not (type(payload['aud']) == str or type(payload['aud']) is list):
+            raise TokenValidationError('Audience (aud) claim must be a string or array of strings present in the ID token')
+
+        if type(payload['aud']) is list and not opt['aud'] in payload['aud']:
+            payload_audiences = ", ".join(payload['aud'])
+            raise TokenValidationError('Audience (aud) claim mismatch in the ID token; expected "{}" but was not one of "{}"'.format(opt['aud'], payload_audiences))
+        elif type(payload['aud']) == str and payload['aud'] != opt['aud']:
+            raise TokenValidationError('Audience (aud) claim mismatch in the ID token; expected "{}" but found "{}"'.format(opt['aud'], payload['aud']))
+
+        # --Time validation (epoch)--
+        now = opt['_clock'] or time.time()
```

Comment:
I'm not assigning this default value at construction time since I don't want to "fix" the value of the clock to whatever returns `time.time()`